File: module_wan/src/main/java/com/huyingbao/module/wan/ui/article/store/ArticleStore.java
package com.huyingbao.module.wan.ui.article.store;

import com.huyingbao.core.arch.dispatcher.RxDispatcher;
import com.huyingbao.core.arch.model.RxAction;
import com.huyingbao.core.arch.model.RxChange;
import com.huyingbao.core.arch.store.RxActivityStore;
import com.huyingbao.module.wan.action.WanResponse;
import com.huyingbao.module.wan.ui.article.action.ArticleAction;
import com.huyingbao.module.wan.ui.article.model.Article;
import com.huyingbao.module.wan.ui.article.model.Banner;
import com.huyingbao.module.wan.ui.article.model.Page;

import org.greenrobot.eventbus.Subscribe;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;
import javax.inject.Singleton;

import androidx.lifecycle.MutableLiveData;

/**
 * Created by liujunfeng on 2019/1/1.
 */
@Singleton
public class ArticleStore extends RxActivityStore {
    private MutableLiveData<List<Article>> mArticleLiveData = new MutableLiveData<>();
    private MutableLiveData<List<Banner>> mBannerLiveData = new MutableLiveData<>();
    private int mNextRequestPage = 1;//列表页数

    @Inject
    ArticleStore(RxDispatcher rxDispatcher) {
        super(rxDispatcher);
    }

    /**
     * 当所有者Activity销毁时,框架调用ViewModel的onCleared（）方法，以便它可以清理资源。
     */
    @Override
    protected void onCleared() {
        super.onCleared();
        mNextRequestPage = 1;
        mArticleLiveData.setValue(null);
        mBannerLiveData.setValue(null);
    }

    /**
     * 接收RxAction
     * 处理RxAction携带的数据
     * 发送RxChange通知RxView
     *
     * @param rxAction
     */
    @Subscribe()
    public void onRxAction(RxAction rxAction) {
        switch (rxAction.getTag()) {
            case ArticleAction.GET_ARTICLE_LIST:
                WanResponse<Page<Article>> articleResponse = rxAction.getResponse();
                if (mArticleLiveData.getValue() == null) {
                    mArticleLiveData.setValue(articleResponse.getData().getDatas());
                } else {
                    mArticleLiveData.getValue().addAll(articleResponse.getData().getDatas());
                    mArticleLiveData.setValue(mArticleLiveData.getValue());
                }
                mNextRequestPage++;
                break;
            case ArticleAction.GET_BANNER_LIST:
                WanResponse<ArrayList<Banner>> bannerResponse = rxAction.getResponse();
                mBannerLiveData.setValue(bannerResponse.getData());
                break;
            case ArticleAction.TO_BANNER:
            case ArticleAction.TO_FRIEND:
            case ArticleAction.TO_LOGIN:
                postChange(RxChange.newInstance(rxAction));
                break;
        }
    }

    public MutableLiveData<List<Article>> getArticleLiveData() {
        return mArticleLiveData;
    }

    public MutableLiveData<List<Banner>> getBannerLiveData() {
        return mBannerLiveData;
    }

    public int getNextRequestPage() {
        return mNextRequestPage;
    }

    public void setNextRequestPage(int nextRequestPage) {
        mNextRequestPage = nextRequestPage;
    }
}

File: module_wan_kotlin/src/main/java/com/huyingbao/module/wan/ui/article/store/ArticleStore.kt
package com.huyingbao.module.wan.ui.article.store

import androidx.lifecycle.MutableLiveData
import com.huyingbao.core.arch.dispatcher.RxDispatcher
import com.huyingbao.core.arch.model.RxAction
import com.huyingbao.core.arch.store.RxActivityStore
import com.huyingbao.module.wan.action.WanResponse
import com.huyingbao.module.wan.ui.article.action.ArticleAction
import com.huyingbao.module.wan.ui.article.model.Article
import com.huyingbao.module.wan.ui.article.model.Banner
import com.huyingbao.module.wan.ui.article.model.Page
import org.greenrobot.eventbus.Subscribe
import java.util.*
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Created by liujunfeng on 2019/1/1.
 */
@Singleton
class ArticleStore @Inject
internal constructor(rxDispatcher: RxDispatcher) : RxActivityStore(rxDispatcher) {
    val articleLiveData = MutableLiveData<ArrayList<Article>>()
    val bannerLiveData = MutableLiveData<ArrayList<Banner>>()
    /**
     * 列表页数
     */
    var nextRequestPage = 1

    /**
     * 当所有者Activity销毁时,框架调用ViewModel的onCleared（）方法，以便它可以清理资源。
     */
    override fun onCleared() {
        super.onCleared()
        nextRequestPage = 1
        articleLiveData.value = null
        bannerLiveData.value = null
    }

    @Subscribe(tags = [ArticleAction.GET_BANNER_LIST])
    fun setBannerLiveData(rxAction: RxAction) {
        val bannerResponse = rxAction.getResponse<WanResponse<ArrayList<Banner>>>()
        bannerLiveData.value = bannerResponse.data
    }

    @Subscribe(tags = [ArticleAction.GET_ARTICLE_LIST])
    fun setArticleLiveData(rxAction: RxAction) {
        val articleResponse = rxAction.getResponse<WanResponse<Page<Article>>>()
        if (articleLiveData.value == null) {
            articleLiveData.setValue(articleResponse.data!!.datas)
        } else {
            articleLiveData.value!!.addAll(articleResponse.data!!.datas!!)
            articleLiveData.setValue(articleLiveData.value)
        }
        nextRequestPage++
    }
}
